### PR TITLE
Add Array#to_proc method [Feature #10829]

### PR DESCRIPTION
**Pull request by richardmarbach**
https://bugs.ruby-lang.org/issues/10829

### CODE REVIEW

File: array.c [hunk 1]
```diff
@@ -2981,6 +2981,40 @@ rb_ary_to_ary_m(VALUE ary)
...
+ *
+ *  If the +array+ is empty, raises ArgumentError.
+ *  If the first element of the +array+ is not a Symbol or a String, raises TypeError.
+ */
+
+
+static VALUE
+ary_proc_call(RB_BLOCK_CALL_FUNC_ARGLIST(recv, ary))
+{
+    argc = (int)RARRAY_LEN(ary);
```
Comment (nobu) [inline array.c:3004]:
```suggestion
    argc = RARRAY_LENINT(ary);
```

File: array.c [hunk 2]
```diff
@@ -2981,6 +2981,40 @@ rb_ary_to_ary_m(VALUE ary)
...
+
+
+static VALUE
+ary_proc_call(RB_BLOCK_CALL_FUNC_ARGLIST(recv, ary))
+{
+    argc = RARRAY_LENINT(ary);
+    rb_check_arity(argc, 1, UNLIMITED_ARGUMENTS);
+    VALUE method = rb_to_symbol(RARRAY_AREF(ary, 0));
+    VALUE rest = rb_ary_subseq(ary, 1, argc);
+    return rb_funcallv(recv, SYM2ID(method), argc - 1, RARRAY_PTR(rest));
```
Comment (nobu) [inline array.c:3008]: Not only this `rest` may be optimized away and garbage-collected, it is unnecessary.
```suggestion
    return rb_funcallv(recv, SYM2ID(method), argc - 1, RARRAY_PTR(ary) + 1);
```

**Comment by nobu** [comment]
IIRC, `Array#to_proc` is an idea proposed during the discussion before numbered parameters, and had been rejected.

**Comment by richardmarbach** [comment]
Thanks for taking the time to review! I learnt a lot working on this :)

Seems like this information was lost in the issue. Feel free to close the PR if the feature has been rejected.

**Comment by nobu** [comment]
> Seems like this information was lost in the issue.
As there were many relevant or similar issues, not all were taken care of.

**Pull request closed without merging**